src/models/StyleTags.js

```diff
@@ -0,0 +1,399 @@
+  const markers = Object.create(null)
+  const sizes = []
+  const names = []
+
+  const insertMarker = id => {
+    const prev = markers[id]
+    if (prev !== undefined) {
+      return prev
+    }
+
+    const marker = (markers[id] = sizes.length)
+    sizes.push(0)
+    return marker
+  }
+
+  const insertRules = (id, cssRules) => {
+    const marker = insertMarker(id)
+    const sheet = sheetForTag(el)
+    const insertIndex = addUpUntilIndex(sizes, marker)
+    sizes[marker] += safeInsertRules(sheet, cssRules, insertIndex)
+  }
+
+  const css = () => {
+    const { cssRules } = sheetForTag(el)
+    let str = ''
+    let i = 0
+
+    // eslint-disable-next-line guard-for-in
+    for (const id in markers) {
+      str += makeTextMarker(id)
+      const end = markers[id] + i
+      for (; i < end; i += 1) {
+        str += cssRules[i].cssText
+      }
+    }
+
+    return str
+  }
+
+  return {
+    styleTag: el,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone: throwCloneTagErr,
+  }
+}
+
+const makeBrowserTag = (el: HTMLStyleElement): Tag<Text> => {
+  const markers = Object.create(null)
+  const names = []
+
+  const insertMarker = id => {
+    const prev = markers[id]
+    if (prev !== undefined) {
+      return prev
+    }
+
+    const marker = (markers[id] = document.createTextNode(makeTextMarker(id)))
+    el.appendChild(marker)
+    return marker
+  }
+
+  const insertRules = (id, cssRules) => {
+    insertMarker(id).appendData(cssRules.join(' '))
+  }
+
+  const css = () => {
+    let str = ''
+    // eslint-disable-next-line guard-for-in
+    for (const id in markers) {
+      str += markers[id].data
+    }
+    return str
+  }
+
+  return {
+    styleTag: el,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone: throwCloneTagErr,
+  }
+}
+
+const makeServerTag = (): Tag<[string]> => {
+  const markers = Object.create(null)
+  const names = []
+
+  const insertMarker = id => {
+    const prev = markers[id]
+    if (prev !== undefined) {
+      return prev
+    }
+
+    return (markers[id] = [makeTextMarker(id)])
+  }
+
+  const insertRules = (id, cssRules) => {
+    const marker = insertMarker(id)
+    marker[0] += cssRules.join(' ')
+  }
+
+  const css = () => {
+    let str = ''
+    // eslint-disable-next-line guard-for-in
+    for (const id in markers) {
+      str += markers[id][0]
+    }
+    return str
+  }
+
+  const tag = {
+    styleTag: null,
+    getIds: getIdsFromMarkersFactory(markers),
+    names,
+    insertMarker,
+    insertRules,
+    css,
+    toHTML: wrapAsHtmlTag(css, names),
+    toElement: wrapAsElement(css, names),
+    clone() {
+      return {
+        ...tag,
+        names: [...names],
```

Comment:
nit: `names.slice()` might be slightly faster and less bytes

Reply:
Yea, I thought that when I was writing this as I copied it over. Not sure what it transpiles to, but good idea, I’ll replace it